### PR TITLE
feature/better-dev-stack-management

### PR DESCRIPTION
**Pull request by JacksonMaxfield**
**Pull request recommendations:**
- [x] Name your pull request _your-development-type/short-description_. Ex: _feature/read-tiff-files_
- [x] Link to any relevant issue in the PR description. Ex: _Resolves [gh-12], adds tiff file format support_
- [x] Provide context of changes.

This has been a long standing effort on my part that was originally started during my attempts to get setup on running pipelines against dev infra. It didn't go well at all. So I decided to make it much easier to setup and manage dev stacks as well as run variations of the event gather pipeline. Should massively help during testing.

Major additions in this PR:
* in the `dev-infrastructure` dir you can now run:
    * `make login` to login to pulumi and gcloud
    * `make init project={project-name}` to generate a whole new gcloud project and pulumi stack
    * `make build` to build and setup the infra
    * `make reset project={project-name}` to clean the infra
    * `make destroy project={project-name}` to teardown and delete all infra
    * `make gen-key project={project-name}` to generate a service account JSON key and store in `.keys/` dir
* added @tohuynh's random event generation function to the repo (will update in example repo accordingly later)
* in the top level dir of the repo you can now run:
    * `make run-min-event-pipeline key=.keys/{key-name.json}` to run the event pipeline with minimal event defition
    * `make run-rand-event-pipeline key=.keys/{key-name.json}` to run the event pipeline with random event defition

- [x] Provide relevant tests for your feature or bug fix.
- [x] Provide or update documentation for any feature added by your pull request.

Documentation upgraded across the board where it touches the additions mentioned.

Thanks for contributing!

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -121,3 +121,10 @@ cdp_database_diagram.*
 cdp_event_gather_flow_*
 ingestion_models.md
 transcript_model.md
+test.err
+test.out
+
+# Common credentials directories
+.keys/
+.credentials/
+.creds/
```
Comment (isaacna) [inline .gitignore:130]: Just curious but is there a reason for both of these?

File: Makefile [hunk 2]
```diff
@@ -46,21 +46,39 @@ clean:  ## clean all build, python, and testing files
...
+docs: ## Generate Sphinx HTML documentation, including API docs, and serve to browser
 	make gen-docs
 	$(BROWSER) docs/_build/html/index.html
+
+run-rand-event-pipeline: ## Run event pipeline using random event gen, requires "key"
+	run_cdp_event_gather \
+		-g $(key) \
+		-e cdp_backend.pipeline.mock_get_events.get_events
+
+run-min-event-pipeline: ## Run event pipeline using minimal event def, requires "key"
```
Comment (isaacna) [inline Makefile:81]: These are both awesome additions!

File: cdp_backend/tests/pipeline/test_event_gather_pipeline.py [hunk 3]
```diff
@@ -14,17 +14,19 @@
...
 
-def mock_get_events_func() -> List[EventIngestionModel]:
+def min_get_events() -> List[EventIngestionModel]:
     event = EXAMPLE_MINIMAL_EVENT
     event.sessions[0].session_datetime = datetime(2019, 4, 13)
     return [event]
 
 
-def test_create_event_gather_flow() -> None:
+@pytest.mark.parametrize("func", [min_get_events, rand_get_events])
```
Comment (isaacna) [inline cdp_backend/tests/pipeline/test_event_gather_pipeline.py:26]: Adding the random test data is a good call

**Review by isaacna — APPROVED** [review]
These are great additions! The new dev commands (especially the key gen and pipeline commands) are gonna be really useful. And the new README is a great resource.

File: .gitignore (hunk 1 above)
Comment (JacksonMaxfield) [reply .gitignore:130]: I have seen people use all of these before and I just want to be safe. Just trying to protect against accidently committing keys.

**Comment by JacksonMaxfield** [comment]
There are two more additions I am going to add to this PR tomorrow / this weekend hopefully:

1. Cleanup / update the base README
2. Add a `make clean` to the dev infrastructure makefile that will simply delete all collections and files in the file store but not hard reset the infra. Which is even better practice than `make reset`.

**Pull request merged**